trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/remote/VirtualHostStyleRemoteS3Facade.java

```diff
@@ -28,9 +28,11 @@ public class VirtualHostStyleRemoteS3Facade
     private final boolean https;
     private final Optional<Integer> port;
 
-    public VirtualHostStyleRemoteS3Facade()
+    @Inject
+    public VirtualHostStyleRemoteS3Facade(RemoteS3Config remoteS3Config)
     {
-        this(VIRTUAL_HOST_BUILDER, true, Optional.empty());
+        this((bucket, region) ->
+                (bucket.isEmpty() ? "" : "%s.".formatted(bucket)) + "s3.%s.%s".formatted(region, remoteS3Config.getDomain()), remoteS3Config.getHttps(), remoteS3Config.getPort());
```

Comment:
One thing that could be interesting is to template the domain part of the config, so that it is not tied to AWS. The current implementation in this PR allows you to change the domain, but still assumes other systems will use the same `<bucket>.s3.<region>.<domain>` format which is not necessarily true.

Could we instead template this with 2 named arguments:
- `%(region)`
- `%(bucket)`

And have 2 configurations:
- `rawDomainTemplate` (for cases without a bucket) - the AWS S3 one would be `s3.%(region).amazonaws.com`
- `bucketDomainTemplate` (for cases with a bucket) - the AWS S3 one would be `%(bucket).s3.%(region).amazonaws.com`

We could then infer whether to load the PathStyle implementation or the VirtualHostStyle one based on whether we have both configs or only `rawDomainTemplate`

NB: I'm using `%(whatever)` encoding for placeholders, that's just the first thing that came to mind - whatever library we want to use for this may use a different format

Reply:
I've added the option of configuring the template option. But, I've removed the case where the bucket is empty. Would there be a scenario where the bucket is empty?

Reply:
Sorry, I missed this comment. Yes, there would be - for instance, listing buckets https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListBuckets.html

Reply:
Added logic for empty buckets